lib/rucio/client/downloadclient.py

```diff
@@ -575,6 +585,7 @@ def _download_item(self, item, trace, log_prefix=''):
                 if not success:
                     logger.warning('%sDownload attempt failed. Try %s/%s' % (log_prefix, attempt, retries))
                     self._send_trace(trace)
+                    self._update_trace(traces_copy_out, trace)
```

Comment:
This would add multiple references to the same trace object. Maybe not a big issue but remember that all traces will only have the highest attempt number.

Reply:
right, sorry, I forgot to remove this one.

Reply:
How about just adding this to the top of download_item:
```
if traces_copy_out is not None:
   traces_copy_out.append(trace)
```

Reply:
and removing all the ` self._update_trace(traces_copy_out, trace)` of course

Reply:
I'll think of it. I wanted to implement it in absolutely the same way, as _send_trace.

Reply:
But _send_trace works different :man_shrugging:

Reply:
the intention was, that whenever downloadclient would send the trace, it attaches the trace to the reference list instead (in case that such reference is given).

Reply:
Right, it creates new trace reference with deepcopy in download_worker. But download worker still works with bulk of files (via queues). So effectively we would add always the same reference in download_item. I think we need to make another deepcopy when updating the trace. But this couldn't be done only at the beginning of the download_item, since we are updating the trace later.

Reply:
actualy download_worker does it within the while loop. So no need for deepcopy. So right, the solution to place it just infront of the download_item should work.